### DISCUSSION TITLE
OIDCSessionMaxDuration set to 0

### DISCUSSION THREAD

**Discussion by logidru** [OpenIDC/mod_auth_openidc#778] [Q&A]
I have set

OIDCSessionMaxDuration=0

so the session will expire when the id token expires. This seems to work as my id token is valid for 5 min and after 5 mins the log says

[Tue Jan 25 15:17:52.075084 2022] [auth_openidc:warn] [pid 11] [client 10.0.0.2:50353] oidc_check_max_session_duration: maximum session duration exceeded for user: SOMEUSERNAME/, referer: 

and the following log is displayed below

[Tue Jan 25 15:18:52.801965 2022] [authz_core:debug] [pid 25] mod_authz_core.c(815): [client 10.0.0.2:50389] AH01626: authorization result of Require valid-user : denied (no authenticated user yet).

But i still can continue after that to use the protected application. i Would expect that i have to relogin, when the session is expired. Did i miss some Config related to session expiration?

When i do then a full reload in the browser a new cookie is created and a new id_token is set in the env vars...

**Comment by zandbelt** [comment]
no additional config is needed; can you provide the (full) server debug logs that show this behaviour?

**Reply by logidru** [reply]
[logs_config_and_comments.zip](https://github.com/zmartzone/mod_auth_openidc/files/7939369/logs_config_and_comments.zip)

Logged in, sucesfully got the id_token and started workiking with the app

id_token iat 06:59:03 
id_token exp 07:04:03


then on 07:04:06 this appears in the log, this seems to be correct, but i can anyway continue to work without problems...

[Wed Jan 26 06:04:06.042333 2022] [auth_openidc:warn] [pid 18] [client 10.0.0.2:51650] oidc_check_max_session_duration: maximum session duration exceeded for user: fa632a18-71ce-4369-9675-b0fbd375f5cd@http://192.168.124.5:8888/auth/realms/Logismata%20OIDC, referer: 


Then on 07:10:00 is press the refresh button in the browser and everything looks normal in the logs. The refresh in the browser somehow creates a new session (new cookie) but i dont need to login again


Note: the server logs display one hour off 


Used Keycloak as IDP (but the same happens when i use auth0)


I noticed that behaviour, when i was triyng to get the refresh token mechanism running, so when i set additionally

OIDCRefreshAccessTokenBeforeExpiry 120

then every 3 Minutes all tokens are refreshed, but i expected that without token refresh the app should not be accessible after the session exeeded.


thx for your help

**Reply by logidru** [reply]
used version
ENV MOD_OPENIDC_VERSION=2.4.10

**Reply by zandbelt** [reply]
thanks, I'll need to review your config as well; it looks like you've used `OIDCUnAuthAction pass`?

Also, `<RequireAny>` was problematic before and has been fixed on master, would. you be able to test from source or would you require a build for a particular platform?

**Reply by logidru** [reply]
Just for explanation

We have installed the module on our reverseproxy and two containers running behind it. The client needs to be protected, but we do the login on the server, therefore we had to protect the server (/viper) as well in order to be able to send the id_token to the server. We added the unauth action to the server because there are some calls that should not require authentication like /viper/serverstatus. As i remember the without OIDCUnAuthAction pass set the behaviour was kind of similar. i guess there was some kind of error message on the client then, that the server was not accessible, but the client was still accessible and after refresh the cookies got renewed as it is now and you could continue to work. I will just remove the OIDCUnAuthAction pass now and test how it behaves.




   <Location /client>
           AuthType openid-connect
           Require valid-user
           LogLevel debug
   </Location>


   <Location /viper>
           RequestHeader set Authorization "Bearer %{OIDC_id_token}e" env=OIDC_id_token
           AuthType openid-connect
           Require valid-user
           LogLevel debug
           OIDCUnAuthAction pass
   </Location>

**Reply by logidru** [reply]
we do our own build directly from the sources into the httpd-alpine image

**Reply by zandbelt** [reply]
as `OIDCUnAuthAction pass` means that users do not have to authenticate, the behaviour makes sense, right? (claims are removed but I do think that the remote user identifier still needs to be cleared correctly)

**Reply by logidru** [reply]
yes, the user needs to authticate on the client to get an id token. This token is used to login on the server. The server does not need to be protected at all. We just needed a way to pass the id_token to the server if there is any.

**Reply by zandbelt** [reply]
so that works right? because the `id_token` is no longer passed after the session expires

**Reply by logidru** [reply]
well not sure if it works...After the session expires, all calls from client to the server fail, it says unauthorized 401 in the client console. But the client is still accessible. And again, after a browser refresh, everything is working again. No relogin required.
[logs.zip](https://github.com/zmartzone/mod_auth_openidc/files/7940658/logs.zip)

**Reply by zandbelt** [reply]
the browser refresh sends the user via Keycloack where an SSO session still exists

**Reply by logidru** [reply]
ok, that makes sense. Just to make sure i understood correctly:

-User calls the app and is redirected the the login screen, logs in and is redirect to the app.
 -> now the user is logged in on keycloak
 -> and is also logged in to the app, id_token valid for 5mins, OIDCSessionMaxDuration=0
-User is working in the app...
-After 5mins the oidc session expires, but the user is still logged in to keycloak
 -> the server returns unauthorized, when it is accessed from the client
 -> the client is still accessible
 
 -> So now i am somehow logged out from the server, but somehow not from the client but in general logged out from mod_auth_openidc
 -> But i am still logged in on keycloak side
 
 -> a Browser reload fixes that issue, but is ugly (user probaly gets a new tomcat session and maybe looses unsaved stuff)
 
 
 ->So in general i have a problem if the session on mod_auth_openidc is expired but the session on the IDP is still intact?
 
 ->One solution would be to set  OIDCRefreshAccessTokenBeforeExpiry 60, so the id and access tokens were renewed before expiration (in my case id and accesstoken have a lifetime of 5min)

**Reply by logidru** [reply]
Or in other words what is the expected behaviour when the mod_auth_openidc session expires (by max_duration timeout or inactivity timeout), when the user is still logged in on the IDP.
And what is the expected behaviour when mod_auth_openidc session expires and the user is NOT logged in anymore on the IDP -> redirect to login screen?

**Reply by logidru** [reply]
just noticed that if the mod_auth_openidc session expires and the user is not logged in on IDP anymore(session deletion via admin gui), the the behaviour is the same. I need to do a reload in the browser order to get the loginscreen to be displayed

**Reply by zandbelt** [reply]
> -> a Browser reload fixes that issue, but is ugly (user probaly gets a new tomcat session and maybe looses unsaved stuff)

setting a longer session timeout seems appropriate for your use case

> ->So in general i have a problem if the session on mod_auth_openidc is expired but the session on the IDP is still intact?

well, that's how SSO is supposed to work; you may want to look into Single Logout

> ->One solution would be to set OIDCRefreshAccessTokenBeforeExpiry 60, so the id and access tokens were renewed before expiration (in my case id and accesstoken have a lifetime of 5min)

that or a longer timeout seems to be appropriate

**Reply by zandbelt** [reply]
> Or in other words what is the expected behaviour when the mod_auth_openidc session expires (by max_duration timeout or inactivity timeout), when the user is still logged in on the IDP. And what is the expected behaviour when mod_auth_openidc session expires and the user is NOT logged in anymore on the IDP -> redirect to login screen?

yes

**Reply by logidru** [reply]
Thx for your quick responses
ok, so i need to take care that the mod_auth_openidc session does not expire, but if it does then i need to somehow manually reconnect to the IDP by refreshing the browser window. With the setting where the server is protected this will work as the app becomes unusable as the server replies unauthorized. But if i allow the server unprotected (OIDCUnAuthAction pass) then the server will never return unauthorized and the user can keep working in the app when the mod_auth_openidc session is already expired. I cannot protect the server just to force the app to be unusable when the session expired. So why is the client app still working even if the session is expired?

**Reply by zandbelt** [reply]
> So why is the client app still working even if the session is expired?

because you've configured `OIDCUnAuthAction pass`

**Reply by logidru** [reply]
So this OIDCUnAuthAction pass is not per Location, as i configured it only for the 

/server location but not for
/client location

**Reply by zandbelt** [reply]
it is per location (that's also why a refresh would trigger full auth)

**Reply by logidru** [reply]
Is there a possibility to get your support on a per hour basis?
